webex_assistant_sdk/crypto.py

```diff
@@ -1,46 +1,106 @@
-        message (str): The message to be signed.
-    Returns:
-        str: The signature. If the secret or message is None return None.
-    """
-    if not (secret and message):
-        raise SignatureGenerationError('The secret or message is not valid.')
-
-    h: hmac.HMAC = _generate_hmac(secret, message)
-    return h.finalize().hex()
-
-
-def verify_signature(secret: str, message: str, signature: str) -> bool:
-    """Verifies a message authentication code
-    Args:
-        secret (str): The secret used to verify the signature.
-        message (str): The message which was signed.
-        signature (str): The signature to verify.
-    Returns:
-        bool: True if the signature is valid and not None, False otherwise
-    """
-    if not signature:
-        return False
-
-    h: hmac.HMAC = _generate_hmac(secret, message)
+import base64
+import binascii
+from pathlib import Path
+from typing import cast
+
+from cryptography.exceptions import UnsupportedAlgorithm
+from cryptography.fernet import Fernet
+from cryptography.hazmat.primitives import hashes, hmac, serialization
+from cryptography.hazmat.primitives.asymmetric.padding import MGF1, OAEP
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey, RSAPublicKey
+from cryptography.hazmat.primitives.serialization import load_ssh_private_key
+
+from webex_assistant_sdk.exceptions import EncryptionKeyError
+
+
+def decrypt(private_key: RSAPrivateKey, message: str) -> str:
+    padding = OAEP(mgf=MGF1(algorithm=hashes.SHA256()),
+                   algorithm=hashes.SHA256(), label=None)
+
+    encrypted_fernet_key, fernet_token = message.split(".")
+    encrypted_fernet_key_bytes = base64.b64decode(
+        encrypted_fernet_key.encode("utf-8"))
+
+    fernet_key = private_key.decrypt(encrypted_fernet_key_bytes, padding)
+
+    fernet_token_bytes = base64.b64decode(fernet_token)
+    payload = Fernet(fernet_key).decrypt(fernet_token_bytes)
+    return payload.decode("utf-8")
+
+
+def verify_signature(secret: str, message: bytes, signature: bytes) -> None:
+    secret_bytes = secret.encode("utf-8")
+
+    sig = hmac.HMAC(secret_bytes, hashes.SHA256())
+    sig.update(message)
+    sig.verify(signature)
+
+
+def load_private_key(private_key_bytes: bytes):
+    """Loads a private key in PEM format"""
     try:
-        h.verify(bytes.fromhex(signature))
-    except InvalidSignature:
-        return False
+        private_key: RSAPrivateKey = load_ssh_private_key(private_key_bytes, None)
+        return private_key
+    except (binascii.Error, ValueError, UnsupportedAlgorithm) as ex:
+        raise EncryptionKeyError('Unable to load private key') from ex
+
+
+def load_private_key_from_file(filename: str):
+    key_data = Path(filename).read_bytes()
+    private_key = load_private_key(key_data)
+    return private_key
+
+
+def load_public_key_from_file(filename: str) -> str:
+    key_data = Path(filename).read_bytes()
+    return key_data.decode('utf-8')
```

Comment:
I feel like it makes more sense if the load public/private key methods return the same types, either both return the bytes or both return the corresponding cryptography object.

Reply:
Done